### PR TITLE
[WIP] GotoDefinition Generates Metadata Signature File For External Symbols

### PR DESCRIPTION
**Pull request by cloudRoutine**
Only one readonly window will open to show signature metadata, I tried a bunch of different ways to make it a preview/provisional texteditor, but nothing panned out.

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
A lot of code looks like reimplementation of layout + niceprint. It would be great to use nicerpint here, after some enhancements, I guess.

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs [hunk 1]
```diff
@@ -30,74 +45,99 @@ type internal FSharpNavigableItem(document: Document, textSpan: TextSpan) =
...
-    [<ImportingConstructor>]
-    (
-        checkerProvider: FSharpCheckerProvider,
-        projectInfoManager: ProjectInfoManager,
-        [<ImportMany>]presenters: IEnumerable<INavigableItemsPresenter>
+type internal FSharpGoToDefinitionService [<ImportingConstructor>]
+    (   checkerProvider: FSharpCheckerProvider
+    ,   projectInfoManager: ProjectInfoManager
+    ,   signatureMetadataService : NavigateToSignatureMetadataService
+    ,   [<ImportMany>]_presenters: IEnumerable<INavigableItemsPresenter>
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: Haskell style detected.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+open Microsoft.CodeAnalysis.Text
+open System.ComponentModel.Composition
+open Microsoft.VisualStudio.FSharp.Editor
+open Microsoft.CodeAnalysis.Formatting
+open Microsoft.VisualStudio.ComponentModelHost
+open Microsoft.VisualStudio.Editor
+open Microsoft.VisualStudio.TextManager.Interop
+open Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem 
+
+[<Export>][<Shared>]
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: I believe `Shared` is by default, you can remove it.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+type internal NavigateToSignatureMetadataService [<ImportingConstructor>] 
+    (   checkerProvider: FSharpCheckerProvider
+    ,   projectInfoManager: ProjectInfoManager
+    ,   workspace : VisualStudioWorkspaceImpl
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: Haskell.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    let mutable currentWindow: IVsWindowFrame option = None 
+
+    // Close the signature metadata window with the solution
+    do workspace.WorkspaceChanged.Add (fun workspaceChange ->
+       match workspaceChange.Kind with 
+       | WorkspaceChangeKind.SolutionCleared
+       | WorkspaceChangeKind.SolutionReloaded
+       | WorkspaceChangeKind.SolutionRemoved ->
+            currentWindow
+            |> Option.bind Option.ofNull
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: I think we should use `Option.ofObj` form FSharp.Core.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    /// If the XML comment starts with '<' not counting whitespace then treat it as a literal XML comment.
+    /// Otherwise, escape it and surround it with <summary></summary>
+    let processXml (xml: string) =
+        if String.IsNullOrEmpty xml then xml else
+        let trimmedXml = xml.TrimStart([|' ';'\r';'\n'|])
+        if String.IsNullOrEmpty trimmedXml then xml else
+        match trimmedXml.[0] with
+        | '<' ->    String.Join ("", "<root>", xml, "</root>")
+        | _   ->    let escapedXml = SecurityElement.Escape xml
+                    String.Join ("", "<summary>", escapedXml, "</summary>")
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: Could you indent branches? It's hard to follow. Maybe an active pattern for `String.IsNullOrEmpty` would work better here.

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: Why not just `"<summary>" + escapedXml + "</summary>"` or `sprintf "<summary>%s</summary>" escapedXml`?

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+                        tryGetFullyQualifiedName mem.EnclosingEntity
+                        |> Option.map (fun parent -> sprintf "%s.%s" parent mem.DisplayName)
+
+                    | TypedAstPatterns.Field (field, _) ->
+                        tryGetFullyQualifiedName field.DeclaringEntity
+                        |> Option.map (fun parent -> sprintf "%s.%s" parent field.DisplayName)
+
+                    | TypedAstPatterns.UnionCase uc ->
+                        match uc.ReturnType with
+                        | TypedAstPatterns.TypeWithDefinition entity ->
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: Inline this match right to `| TypedAstPatterns.UnionCase uc `?

File: vsintegration/tests/unittests/Tests.OpenDeclarationsGetter.fs [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    let decls =
+        [ { openDecl [ { openDeclWithAutoOpens ["Other.Already.Matched"] with IsUsed = true }] with IsUsed = true }
+          openDecl [ openDeclWithAutoOpens ["System.IO"]]
+        ]
+
+    let updatedDecls = 
+        OpenDeclarationGetter.updateOpenDeclsWithSymbolPrefix 
+            [|"System"; "IO"|] (range (1, 1) (2, 2)) decls
+
+    CollectionAssert.AreEqual ([true; true], updatedDecls |> List.map (fun decl -> decl.IsUsed))
```
Comment (vasily-kirichenko) [inline vsintegration/tests/unittests/Tests.OpenDeclarationsGetter.fs]: Use defined above `Collection.assertEqual` or delete that and similar helpers.

File: vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs [hunk 8]
```diff
@@ -882,4 +718,951 @@ module internal SymbolUse =
...
+        | ".ctor" | ".cctor" -> Some func.EnclosingEntity
+        | _ -> None
+
+    let (|Function|_|) excluded (func: FSharpMemberOrFunctionOrValue) =
+        match func.FullTypeSafe |> Option.map getAbbreviatedType with
+        | Some typ when typ.IsFunctionType
+                       && not func.IsPropertyGetterMethod 
+                       && not func.IsPropertySetterMethod
+                       && not excluded
+                       && not (isOperator func.DisplayName) -> Some()
```
Comment (smoothdeveloper) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: @vasily-kirichenko here also, haskell police :)

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs (hunk 1 above)
Comment (smoothdeveloper) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: We should keep any style which doesn't look unkept and look like some work was put to layout the code for better reading.

Trying to harmonize the codebase at large to a single style doesn't seem bring as much as just making sure code has had efforts to make readable.

I prefer this (by far) versus all on single or two lines.

File: vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs [hunk 9]
```diff
@@ -882,4 +718,951 @@ module internal SymbolUse =
...
+                           Some() 
+                       | _ -> 
+                           None
+                   res)))
+         |> Option.isSome
+
+    let isOperator (name: string) =
+        name.StartsWith "( " && name.EndsWith " )" && name.Length > 4
+            && name.Substring (2, name.Length - 4) 
+               |> String.forall (fun c -> c <> ' ' && not (Char.IsLetter c))
```
Comment (smoothdeveloper) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: suggestions:

* maybe this whole expression can be better laid out?
* also could you have comment on stuff dealing with substring as those are really annoying to reverse engineer / run in the debugger in my head.

proposal:

```fsharp
   let isOperator (name: string) =
        let inline notSpaceOrLetter c = c <> ' ' && not (Char.IsLetter c)
        name.StartsWith "( "
        && name.EndsWith " )"
        && name.Length > 4
        // check something something
        && (name.Substring (2, name.Length - 4) |> String.forall notSpaceOrLetter) 
```

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs (hunk 5 above)
Comment (smoothdeveloper) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: @vasily-kirichenko I've seen this style of indentation when first branch is simple, and the code will need further cases added, I think it makes sense.

See https://github.com/smoothdeveloper/visualfsharp/blob/a3a3fb552d27ad8ba5cab6497a033ebf19655025/src/fsharp/NicePrint.fs#L1965-L2011 and try to indent the stuff, as you'll add cases, you'll need to go deeper on indentation and it will look really odd.

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs (hunk 1 above)
Comment (cloudRoutine) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: most of the time i end up writing this way because i'm commenting and switching out services in the importing constructor, trying to figure out which one will actually do the right thing. it's more practical than stylistic

**Comment by Pilchie** [comment]
@jasonmalinowski or @dpoeschl - can one of you take a look at this?

File: vsintegration/src/FSharp.Editor/Common/LanguageService.fs [hunk 10]
```diff
@@ -193,52 +200,54 @@ type internal FSharpCheckerWorkspaceServiceFactory
...
-                             DefaultToNonHotURLs = true,
-                             EnableCommenting = true,
-                             CodeSenseDelay = 100,
-                             ShowDropDownOptions = true)>]
-    internal FSharpPackage() =
+
+[<Guid(FSharpCommonConstants.packageGuidString)>]
+[<ProvideLanguageService 
+    (   languageService = typeof<FSharpLanguageService>
+    ,   strLanguageName = FSharpCommonConstants.FSharpLanguageName
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/LanguageService.fs]: Please no commas at start of line. Please stick to the coding conventions already in use in the codebase

File: vsintegration/src/FSharp.Editor/Common/LanguageService.fs [hunk 11]
```diff
@@ -193,52 +200,54 @@ type internal FSharpCheckerWorkspaceServiceFactory
...
-    [<ProvideEditorExtension(FSharpCommonConstants.editorFactoryGuidString, ".ml", 97)>]
-    [<ProvideEditorExtension(FSharpCommonConstants.editorFactoryGuidString, ".mli", 97)>]
+
+
+and [<Guid(FSharpCommonConstants.languageServiceGuidString)>]
+    [<ProvideService (typeof<FSharpLanguageService>,ServiceName="F# Language Service")>]
+    [<ProvideLanguageExtension (typeof<FSharpLanguageService>, ".fs")>]
+    [<ProvideLanguageExtension (typeof<FSharpLanguageService>, ".fsi")>]
+    [<ProvideLanguageExtension (typeof<FSharpLanguageService>, ".fsx")>]
+    [<ProvideLanguageExtension (typeof<FSharpLanguageService>, ".fsscript")>]
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/LanguageService.fs]: Please don't reformat code as part of PRs - minimize the diffs please and do reformatting in a separate PR.  Large diffs makes it hard to review.

File: vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+        let matched = decls |> List.exists fst
+        let isUsed = decls |> List.exists (fun (_, decl) -> decl.IsUsed)
+        //if not decl.IsUsed && isUsed then debug "OpenDeclaration %A is used by %A" decl symbolPrefix
+        matched, { decl with Declarations = decls |> List.map snd; IsUsed = isUsed }
+
+module internal OpenDeclarationGetter =
+    open System 
+    open System.Diagnostics
+
+    let private getAutoOpenModules entities =
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs]: Please try to make sure every new top-level function, module and type definition added to the codebase has a `///` comment.  (I'm seeing a huge amount of code flowing into the codebase in recent PRs that isn't following this - so it's a general comment for all contributors)

File: vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+        List.foldBack( fun (e: RawEntity) acc ->
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs]: In the compiler code it is more normal to use ``(as,bs) ||> List.foldBack (fun a b  -> ...)``

File: vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+
+    let private getAutoOpenModules entities =
+        List.foldBack( fun (e: RawEntity) acc -> 
+             match e.Kind LookupType.Precise with
+             |  EntityKind.Module { IsAutoOpen = true } -> e.CleanedIdents :: acc
+             | _ -> acc) entities []
+
+
+    let private getActivePatterns (entities: RawEntity list) =
+        List.foldBack( fun (e: RawEntity) acc ->
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs]: This is a ``List.filter`` or ``List.choose``.  Please don't use ``fold`` operations when a better alternative is available.

File: vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs [hunk 15]
```diff
@@ -0,0 +1,476 @@
+﻿namespace Microsoft.VisualStudio.FSharp.Editor
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs]: I don't really know what this file is doing.  My suspicion is that it should be part of the FCS API rather than in FSHarp.Editor. Please add a signature file to make the overall implemented operation clearer and allow future migration of the functionality behind the FCS API barrier.  Also I'm concerned that there is duplication here somehow with other operations in either FCS or FSharp.Editor

File: vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+                                    module A2 =
+                                        module A3 = 
+                                            [<AutoOpen>] 
+                                            module A4 = ...
+                                     
+                            // this declaration actually open M, M.A1, M.A1.A2, but NOT M.A1.A2.A3 or M.A1.A2.A3.A4
+                            open M
+                        *)
+
+                        let rec loop' acc maxLength =
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs]: I'm concerned how hard this sort of code is to understand (``let rec loop'`` followed further down by another ``let rec loop``).  The iteration patterns can almost always be abstracted.  At the moment it's really hard to follow,

File: vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs [hunk 17]
```diff
@@ -0,0 +1,476 @@
...
+                            else acc
+                        | SynModuleSigDecl.Open (longIdent, openDeclRange) ->
+                            let identArray = processOpenDeclaration longIdent
+                            if openDeclRange.EndLine < pos.Line || (openDeclRange.EndLine = pos.Line && openDeclRange.EndColumn < pos.Column) then 
+                                String.Join(".", identArray) :: acc
+                            else acc
+                        | _ -> acc) openStatements
+
+                modules
+                |> List.fold (fun acc (SynModuleOrNamespaceSig(_, _, _, decls, _, _, _, moduleRange)) ->
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs]: Please remove nearly all the uses of ``fold`` in this module in favor of other combinators (as mentioned above).  They all look like filters, choose, etc.

File: vsintegration/src/FSharp.Editor/Common/SignatureGenerator.fs [hunk 18]
```diff
@@ -0,0 +1,1055 @@
+﻿module Microsoft.VisualStudio.FSharp.Editor.SignatureGenerator
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/SignatureGenerator.fs]: This code is indeed a major duplication of the NicePrint.fs functionality.  I believe we need to have one and only one implementation of the process of signature printing and formatting, otherwise we are headed for major maintenance problems.

Please work out how to use the NicePrint.fs code instead of this code

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs (hunk 1 above)
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: > We should keep any style which doesn't look unkept and look like some work was put to layout the code for better reading.  Trying to harmonize the codebase at large to a single style doesn't seem bring as much as just making sure code has had efforts to make readable.

The contribution guidelines are clear - use the styles that are already inn use in the compiler.  Please don't add new stylings

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs (hunk 2 above)
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: Please use ``[<Export; Shared>]`` if there are multipe attributes

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs (hunk 5 above)
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSignatureMetadataService.fs]: > @vasily-kirichenko I've seen this style of indentation when first branch is simple, and the code will need further cases added, I think it makes sense.

@smoothdeveloper Here @vasily-kirichenko's is correct that it is normal to indent branches for this kind of code

**Comment by cloudRoutine** [comment]
I'll try rewriting it to use NicePrint instead of the old VFPT approach

**Comment by dsyme** [comment]
> I'll try rewriting it to use NicePrint instead of the old VFPT approach

@cloudRoutine It's tricky because I understand the VFPT code is probably in some ways better than NicePrint.fs.....

**Comment by cloudRoutine** [comment]
@dsyme I could use a little more clarification about how you want me to approach this. Do you want me to possibly add additional functionality to NicePrint and then expose new functions in SourceCodeServices to get the signature for an FSharpSymbol? 

Should I splice the doccoms into the signature after receiving it from the compiler service, or can I add an additional function to provide an "annotated" signature as well?

I've read the [F# Compiler Guide](https://fsharp.github.io/2015/09/29/fsharp-compiler-guide.html), [contributing.md](https://github.com/Microsoft/visualfsharp/blob/master/CONTRIBUTING.md), and have made a few PRs in the past, but it's still fairly unclear to me where certain kinds of functionality are supposed to be implemented, especially when it comes to editor functionality that could work on either side of the FCS boundary.

**Comment by dsyme** [comment]
> @dsyme I could use a little more clarification about how you want me to approach this. Do you want me to possibly add additional functionality to NicePrint and then expose new functions in SourceCodeServices to get the signature for an FSharpSymbol?

It's really hard.  I'll take a closer look at the PR.  Maybe we can live with the duplication

**Comment by cloudRoutine** [comment]
you don't need to spend to much time on it at the moment, i'm just getting back to it right now. 
> expose new functions in SourceCodeServices to get the signature for an FSharpSymbol
^ this is what'll I'll aim for if I can't make it work I'll say so, if I do you can review that

**Comment by dsyme** [comment]
> > expose new functions in SourceCodeServices to get the signature for an FSharpSymbol
> ^ this is what'll I'll aim for if I can't make it work I'll say so, if I do you can review that

OK. This makes sense.  It depends a lot on the quality of the signature generation code being added here.  If people (you and @vasily-kirichenko) believe it to be better than what exists in NicePrint.fs then we can accept the duplication - either in FCS or FSharp.Editor.

If it's worse - or better in some ways and substantially worse in other - then it's less convincing.

Honestly I'll take the advice or you and @vasily-kirichenko about what's the right thing to do here :)

**Comment by cloudRoutine** [comment]
If possible I'd much rather have `SourceCodeServices` be able to provide the signatures even if it takes a lot more effort to bring `NicePrint`, that way other editors can take advantage of it and it could probably also be used to generate signature stubs from an implementation file, getting rid of some of the tedium of working with signatures.

**Comment by dsyme** [comment]
> If possible I'd much rather have SourceCodeServices be able to provide the signatures even if it takes a lot more effort to bring NicePrint, that way other editors can take advantage of it and it could probably also be used to generate signature stubs from an implementation file, getting rid of some of the tedium of working with signatures.

Yes it would be good to have it generally available.

I think we should probably just put the code you already have behind the SourceCodeServices boundary and accept the duplication - as long as you and @vasily-kirichenko feel the code is up to scratch and can be progressed.    It will get tested quite thoroughly quite quickly by actual users (it's a very visible feature) so if there are deficiencies they will become apparent pretty quickly.

**Comment by vasily-kirichenko** [comment]
@KevinRansom I think we should just merge it as is. Nobody is gonna make it "right".

**Comment by dsyme** [comment]
Yes - @cloudRoutine  I'll reopen this if that's ok.  I agree we need to get this into shape in some form (and it might be me who needs to do it since I'm responsible for the awfulness of the NicePrint.fs code)

**Comment by cloudRoutine** [comment]
@dsyme sure np. I closed it since I'm gonna take a hiatus from working on VFT, it's outdated, full of conflicts, and I figured no one else would work on it.

**Comment by realvictorprm** [comment]
@cloudRoutine I will take over.

**Comment by realvictorprm** [comment]
@cartermp let's fix this. I need some advice as soon as I'm into it enough

**Comment by realvictorprm** [comment]
The PR is still working for me. With which should I start?
- Update to latest master
- Fix what was proposed

cc @dsyme, @cartermp

Also I would like to fix this small bug here:
image

All types should be known through the assembly references of the project.

**Comment by dsyme** [comment]
> The PR is still working for me. With which should I start?

Yes, update to latest master.  Submit a new PR if needed (though @cloudRoutine will likely give you push access to his repo if you wish)

**Comment by dsyme** [comment]
> All types should be known through the assembly references of the project.

I think that will be hard to fix.  There is no project available to do this checking

**Comment by cartermp** [comment]
I certainly won't block on this:

> All types should be known through the assembly references of the project.

The benefit of the other features definitely outweighs this issue.

**Comment by realvictorprm** [comment]
@dsyme Can't we reference all dll's of the project we are currently in?

**Comment by dsyme** [comment]
> @dsyme Can't we reference all dll's of the project we are currently in?

The F# Language Service needs a "project" for the .fsi file to live in.  

We do already manufacture a "single file project" for the standalone .fsi file.  We could in theory add the references to this single-file project but we don't do anything like that just ye, and I'm not sure it's that easy to plumb the information across.  If it turns out to be easy then we should do it, but I think it can be separate as it's very different work to the current PR

**Comment by Pilchie** [comment]
> We do already manufacture a "single file project" for the standalone .fsi file,

In Roslyn there is a [`MiscellaneousFilesWorkspace`](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,344ad90895765109), and a [`MetadataAsSourceWorkspace`](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/MetadataAsSource/MetadataAsSourceWorkspace.cs,fd907b0eb777fc19), and in the latter we create projects for the generated source.  It still doesn't solve *all* issues though (because you might not have references to internal interfaces that are implemented, etc).  For that reason Roslyn also disables semantic error checking in those files.

**Comment by realvictorprm** [comment]
My suggestion is to reference at least the assembly where the meta data is from.

**Comment by dsyme** [comment]
> My suggestion is to reference at least the assembly where the meta data is from.

@realvictorprm You are totally right - we should either do this or not show any errors for these generated files. I'm really just saynig that we can tolerate the errors being shown initially and deal with it in a separate PR

**Comment by KevinRansom** [comment]
@dsyme, @cloudRoutine 

Is this still a WIP, or is it something that should be closed?

Thanks

Kevin

**Comment by dsyme** [comment]
Closing very old PR, I think maser will have diverged too far
